File: model/collection.go
package model

import (
	"time"

	"github.com/allape/gocrud"
)

type CollectionType string

const (
	CollectionTypePlaylist CollectionType = "playlist"
	CollectionTypeArtist   CollectionType = "artist"
	CollectionTypeAlbum    CollectionType = "album"
	CollectionTypeLanguage CollectionType = "language"
	CollectionTypeOST      CollectionType = "ost"
	CollectionTypeOpera    CollectionType = "opera"
)

var CollectionTypes = []CollectionType{
	CollectionTypePlaylist,
	CollectionTypeArtist,
	CollectionTypeAlbum,
	CollectionTypeLanguage,
	CollectionTypeOST,
	CollectionTypeOpera,
}

type Collection struct {
	gocrud.Base
	Type        CollectionType `json:"type"`
	Cover       string         `json:"cover"`
	Name        string         `json:"name"`     // human-readable name
	Keywords    string         `json:"keywords"` // search keywords
	Code        string         `json:"code"`     // for machine usage, not for human
	Description string         `json:"description"`
}

type Role string

const (
	Singer   Role = "singer"
	Lyricist Role = "lyricist"
	Composer Role = "composer"
	Arranger Role = "arranger"
	Producer Role = "producer"
	Other    Role = "other"
	Reserved Role = "_"
)

var Roles = []Role{
	Singer,
	Lyricist,
	Composer,
	Arranger,
	Producer,
	Other,
	Reserved,
}

type CollectionSong struct {
	SongID       gocrud.ID `json:"songId" gorm:"index:idx_collection_song"`
	CollectionID gocrud.ID `json:"collectionId" gorm:"index:idx_collection_song"`
	Role         Role      `json:"role" gorm:"default:'_';index:idx_collection_song"`
	CreatedAt    time.Time `json:"createdAt" gorm:"autoCreateTime;<-:create"`
}

File: ui/admin/src/model/collection.ts
import { IBase, IBaseSearchParams } from "@allape/gocrud";
import { IColoredLV } from "@allape/gocrud-react";
import { useMemo } from "react";
import { useTranslation } from "react-i18next";
import { ISong } from "./song.ts";

export type CollectionType =
  "playlist" | "artist" | "album" | "language" | "ost" | "opera";

export interface ICollection extends IBase {
  type: CollectionType;
  cover: string;
  name: string;
  keywords: string;
  code: string;
  description: string;
}

export interface ICollectionSearchParams extends IBaseSearchParams {
  in_type?: ICollection["type"][];
  keywords?: string;
}

export function useCollectionTypes(): IColoredLV<CollectionType>[] {
  const { t } = useTranslation();
  return useMemo<IColoredLV<CollectionType>[]>(
    () =>
      CollectionTypes.map((ct) => ({
        ...ct,
        label: t(ct.label as string),
      })),
    [t],
  );
}

export const CollectionTypes: IColoredLV<CollectionType>[] = [
  {
    label: "collection.types.playlist",
    value: "playlist",
    color: "blue",
  },
  {
    label: "collection.types.artist",
    value: "artist",
    color: "green",
  },
  {
    label: "collection.types.album",
    value: "album",
    color: "orange",
  },
  {
    label: "collection.types.language",
    value: "language",
    color: "yellow",
  },
  {
    label: "collection.types.ost",
    value: "ost",
    color: "cyan",
  },
  {
    label: "collection.types.opera",
    value: "opera",
    color: "lime",
  },
];

export const ArtistCollectionTypes: CollectionType[] = ["artist"];

export const NonArtistCollectionTypes: CollectionType[] =
  CollectionTypes.filter((ct) => !ArtistCollectionTypes.includes(ct.value)).map(
    (ct) => ct.value,
  );

export type Role =
  "singer" | "lyricist" | "composer" | "arranger" | "producer" | "other" | "_";

export interface ICollectionSong extends Pick<IBase, "createdAt"> {
  songId: ISong["id"];
  collectionId: ICollection["id"];
  role: Role;
}

export function FromCollectionIds(
  collectionIds: ICollectionSong["collectionId"][],
  songId: ICollectionSong["songId"],
  role: Role,
): ICollectionSong[] {
  return collectionIds.map(
    (ci) =>
      ({
        collectionId: ci,
        songId,
        role,
      }) as ICollectionSong,
  );
}
